### PR TITLE
fix:`Block 174731` - `ids.a` Expected as Integer

### PR DESCRIPTION
**Pull request by whichqua**
Issue Number: #429

## Type

- [ ] feature
- [x] bugfix
- [ ] dev (no functional changes, no API changes)
- [ ] fmt (formatting, renaming)
- [ ] build
- [ ] docs
- [ ] testing

## Description
The issue is triggered when calling the `get_tx_info()` function inside `__validate_deploy__`.
The steps to fix the issue:
- [x] Replicate the issue with a simple integration test
- [x] Find the root of the issue
- [x] Write a fix
- [x] Sepolia test

## Breaking changes?

- [ ] yes
- [x] no

### CODE REVIEW

File: tests/integration/contracts/blockifier_contracts/feature_contracts/cairo0/test_proxy.cairo [hunk 1]
```diff
@@ -0,0 +1,112 @@
+%lang starknet
+
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+from starkware.starknet.common.syscalls import library_call, library_call_l1_handler
+from starkware.cairo.common.math import assert_not_zero
+
```
Comment (notlesh) [inline tests/integration/contracts/blockifier_contracts/feature_contracts/cairo0/test_proxy.cairo:6]: If there's any context you can add about this file (where it came from, why we have it, etc.) it would be good to add here.

File: tests/integration/contracts/blockifier_contracts/feature_contracts/cairo0/test_proxy_account.cairo [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+} (selector: felt, calldata_size: felt, calldata: felt*) {
+    alloc_locals;
+    // get the tx info
+    let (tx_info) = get_tx_info();
+    with_attr error_message("assertion failed") {
+        assert_nn(tx_info.signature_len - 2);
+    }
+
+    return ();
+}
```
Comment (notlesh) [inline tests/integration/contracts/blockifier_contracts/feature_contracts/cairo0/test_proxy_account.cairo:90]: Is this still needed? If we don't need all of the unusual things here (e.g. @raw_input) in order to test this (as a regression test) then I would suggest removing it or at least adding some documentation about it. Otherwise I can see it adding a lot of confusion in the future.

File: tests/integration/deploy_txn_tests.rs [hunk 3]
```diff
@@ -358,3 +358,144 @@ async fn deploy_via_invoke_no_calldata_cairo1_account(
...
+        initial_state.cairo1_compiled_classes,
+        HashMap::default(),
+    )
+    .await
+    .expect("OS run failed");
+}
+
+#[rstest]
+#[tokio::test(flavor = "multi_thread", worker_threads = 1)]
+async fn deploy_cairo0_via_proxy(block_context: BlockContext, max_fee: Fee) {
```
Comment (notlesh) [inline tests/integration/deploy_txn_tests.rs:429]: Does the proxy impact this bug at all? I might apply my same concerns above if not, this is likely to add confusion in the future.

File: crates/starknet-os/src/hints/mod.rs [hunk 4]
```diff
@@ -626,8 +626,7 @@ pub async fn start_tx_async<PCS>(
 where
     PCS: PerContractStorage + 'static,
 {
-    let deprecated_tx_info_ptr =
-        get_relocatable_from_var_name(vars::ids::DEPRECATED_TX_INFO, vm, ids_data, ap_tracking)?;
+    let deprecated_tx_info_ptr = get_ptr_from_var_name(vars::ids::DEPRECATED_TX_INFO, vm, ids_data, ap_tracking)?;
```
Comment (notlesh) [inline crates/starknet-os/src/hints/mod.rs:629]: image

**Review by notlesh — COMMENTED** [review]
LGTM, but I think we should clean up some of the test code that isn't relevant

File: crates/starknet-os/src/hints/mod.rs (hunk 4 above)
Comment (HermanObst) [reply crates/starknet-os/src/hints/mod.rs:629]: HAHAHAHAHAHAHAHHA ❤️

File: tests/integration/contracts/blockifier_contracts/feature_contracts/cairo0/test_proxy_account.cairo (hunk 2 above)
Comment (whichqua) [reply tests/integration/contracts/blockifier_contracts/feature_contracts/cairo0/test_proxy_account.cairo]: Resolved by https://github.com/keep-starknet-strange/snos/pull/431/commits/c4726a7c44e2a64d7156aeaea7075d52c14ee601

File: tests/integration/deploy_txn_tests.rs (hunk 3 above)
Comment (whichqua) [reply tests/integration/deploy_txn_tests.rs]: Resolved by https://github.com/keep-starknet-strange/snos/pull/431/commits/c4726a7c44e2a64d7156aeaea7075d52c14ee601

File: crates/starknet-os/src/hints/tests.rs [hunk 5]
```diff
@@ -158,7 +158,9 @@ pub mod tests {
         vm.add_memory_segment();
         vm.add_memory_segment();
 
-        let ids_data = ids_data![vars::ids::DEPRECATED_TX_INFO];
+        let mut ids_data = HashMap::new();
+        ids_data.insert(vars::ids::DEPRECATED_TX_INFO.to_owned(), HintReference::new(0, 0, false, false));
+
```
Comment (HermanObst) [inline crates/starknet-os/src/hints/tests.rs:163]: @whichqua why this change is needed?

Comment (whichqua) [reply crates/starknet-os/src/hints/tests.rs:163]: Because the vm treats `pointers` and `relocalables` differently. This is how to define pointers.

File: tests/integration/deploy_txn_tests.rs [hunk 6]
```diff
@@ -358,3 +358,63 @@ async fn deploy_via_invoke_no_calldata_cairo1_account(
...
     .expect("OS run failed");
 }
+
+#[rstest]
+#[tokio::test(flavor = "multi_thread", worker_threads = 1)]
+async fn deploy_cairo0_check_get_info_call(block_context: BlockContext, max_fee: Fee) {
+    let account_with_tx_info_check = load_cairo0_feature_contract("account_with_tx_info_check");
+    let class_hash = class_hash!("0x6c8903651a5f89ffc304621a7d8106a0324cc28aca04934fcbbb4398d5c8bc8");
+
+    let args: Vec<Felt252> = [].into_iter().map(Felt252::from_hex_unchecked).collect();
```
Comment (HermanObst) [inline tests/integration/deploy_txn_tests.rs:368]: Why not just?

`let args: Vec<Felt252> = Vec::new();
`

**Review by HermanObst — CHANGES_REQUESTED** [review]

**Review by HermanObst — APPROVED** [review]
LGTM!

**Review by ftheirs — APPROVED** [review]
This looks awesome 😎
